nav2_pure_pursuit/src/pure_pursuit_controller.cpp

```diff
@@ -0,0 +1,414 @@
+    (carrot_pose.pose.position.y * carrot_pose.pose.position.y);
+
+  // Find curvature of circular arc
+  double curvature = 0.0;
+  if (carrot_dist2 > 0.001) {
+    curvature = 2.0 * carrot_pose.pose.position.y / carrot_dist2;
+  }
+
+  // Apply curvature to angular velocity
+  linear_vel = desired_linear_vel_;
+  angular_vel = desired_linear_vel_ * curvature;
+
+  // Make sure we're in compliance with basic kinematics
+  rclcpp::Duration dt = rclcpp::Time(pose.header.stamp) - rclcpp::Time(last_cmd_.header.stamp);
+  applyKinematicConstraints(
+    linear_vel, angular_vel,
+    fabs(lookahead_dist - sqrt(carrot_dist2)), lookahead_dist, dt.seconds());
+
+  // Collision checking
+  if (isCollisionImminent(pose, carrot_pose)) {
+    RCLCPP_ERROR(logger_, "Collision imminent!");
+    throw std::runtime_error("PurePursuitController detected collision ahead!");
+  }
+
+  // populate and return message
+  geometry_msgs::msg::TwistStamped cmd_vel;
+  cmd_vel.header.frame_id = pose.header.frame_id;
+  cmd_vel.header.stamp = clock_->now();
+  cmd_vel.twist.linear.x = linear_vel;
+  cmd_vel.twist.angular.z = angular_vel;
+  last_cmd_ = cmd_vel;
+  return cmd_vel;
+}
+
+geometry_msgs::msg::PoseStamped PurePursuitController::getLookAheadMarker(
+  const double & lookahead_dist,
+  const nav_msgs::msg::Path & transformed_plan)
+{
+  // Find the first pose which is at a distance greater than the lookahead distance
+  auto goal_pose_it = std::find_if(
+    transformed_plan.poses.begin(), transformed_plan.poses.end(), [&](const auto & ps) {
+      return hypot(ps.pose.position.x, ps.pose.position.y) >= lookahead_dist;
+    });
+
+  // If the no pose is not far enough, take the last pose
+  if (goal_pose_it == transformed_plan.poses.end()) {
+    goal_pose_it = std::prev(transformed_plan.poses.end());
+  }
+
+  return *goal_pose_it;
+}
+
+bool PurePursuitController::isCollisionImminent(
+  const geometry_msgs::msg::PoseStamped & robot_pose,
+  const geometry_msgs::msg::PoseStamped & carrot_pose)
+{
+  // This may be a bit unusual, but the robot_pose is in odom frame
+  // and the carrot_pose is in robot base frame. We need to collision
+  // check in odom frame, so all values will be relative to robot base pose.
+  // But we can still use the carrot pose to find quantities and convert to odom.
+
+  const double carrot_dist = hypot(carrot_pose.pose.position.x, carrot_pose.pose.position.y);
+  const unsigned int num_pts = static_cast<unsigned int>(ceil(carrot_dist / costmap_->getResolution()));
+  const double pt_distance = carrot_dist / num_pts;
+
+  geometry_msgs::msg::PoseStamped carrot_in_odom;
+  try {
+    tf_->transform(carrot_pose, carrot_in_odom,
+      costmap_ros_->getGlobalFrameID(), transform_tolerance_);
+    return true;
+  } catch (tf2::TransformException & ex) {
+    RCLCPP_ERROR(
+      logger_,
+      "Exception attempting to get look ahead point in odom frame: %s", ex.what());
+  }
```

Comment:
We can use the `transformPose` function for this.

Reply:
Good point